### PR TITLE
feat: skip logprob and reference logprob computation under certain conditions

### LINKED ISSUE #1186: Remove unnecessary model forward passes in logprob phase in GRPO
In GRPO we perform a forward on the policy model to get the logprob, and then we perform training; when rollout GBS == training GBS, the first part is not necessary because it's doing the same thing as training's forward pass, and we can extract logprob from the forward's output.

### PR DESCRIPTION
**Pull request by guyueh1**
# What does this PR do ?

Skip logprob and reference logprob computation under certain conditions:
* when `loss_fn.skip_reference_policy_logprobs_calculation=true`, skip reference logprob. The requirement is `loss_fn.reference_kl_penalty == 0` which will be checked whenever `skip_reference_policy_logprobs_calculation` is true.
* when `loss_fn.force_on_policy_ratio=true`, skip logprob computation. The requirement is rollout batch size == train global batch size, which will be checked whenever `force_on_policy_ratio` is true.

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

## Release Notes

* **New Features**
  * Added configuration options to optimize GRPO training: ability to skip reference policy logprob calculations and enforce on-policy ratio in loss function computations.
  * New flags reduce computational overhead during training while maintaining training stability.

### CODE REVIEW

**Review by terrykong — COMMENTED** [review]
@guyueh1 i believe this covers 012af2a51d22def1ff170ecb3594ffe538898ff9, but just double checking

**Review by terrykong — COMMENTED** [review]
@guyueh1 and c759a4eba5d739650e6276ad0d4e3aa03985bb34

**Comment by guyueh1** [comment]
> 012af2a51d22def1ff170ecb3594ffe538898ff9

No, but I think I can add this to my PR

> c759a4eba5d739650e6276ad0d4e3aa03985bb34

Yes

**Comment by guyueh1** [comment]
@youngeunkwon0405 can you help review?

File: nemo_rl/algorithms/grpo.py [hunk 1]
```diff
@@ -1727,11 +1725,26 @@ def grpo_train(
...
+                    )
+                    train_data["prev_logprobs"] = torch.zeros_like(
+                        train_data["generation_logprobs"]
+                    )
+                if not (skip_prev_logprobs and skip_reference_policy_logprobs):
+                    print("▶ Preparing for logprob inference...", flush=True)
+                    with timer.time("logprob_inference_prep"):
+                        policy.prepare_for_lp_inference()
 
+                memory_tracker.snapshot_start_of_stage("Computing logprobs", dir())
```
Comment (youngeunkwon0405) [inline nemo_rl/algorithms/grpo.py]: Before, it was taking a snapshot before doing offloading. In this PR, it is taking the snapshot after the offloading. Is this an intentional change?

**Review by youngeunkwon0405 — COMMENTED** [review]
Hi @guyueh1, should ' force_on_policy_ratio=True ' be used only when `pps * gpp == training.batch_size`?
If so, I think we need an assertion to force that condition.

+ I think `force_on_policy_ratio` should be automatically turned on when `pps * gpp == training.batch_size`. What do you think?

**Comment by terrykong** [comment]
@yfw @guyueh1 can we close out #2177 / #2178 and #2174 and just use this one?

**Comment by guyueh1** [comment]
> Hi @guyueh1, should ' force_on_policy_ratio=True ' be used only when `pps * gpp == training.batch_size`? If so, I think we need an assertion to force that condition.
> 
> * I think `force_on_policy_ratio` should be automatically turned on when `pps * gpp == training.batch_size`. What do you think?

@youngeunkwon0405 there is already an assertion in main branch to make sure pps*gpp == training_gbs : https://github.com/NVIDIA-NeMo/RL/blob/cc75cadfe061301bd121306f1648957583760528/nemo_rl/algorithms/grpo.py#L381-L390

the flag `force_on_policy_ratio` was not added by this PR, i think the idea, when it was first added, is to avoid changing the previous behavior unless user explicitly do so. I personally don't want to delete it in this PR.

File: nemo_rl/algorithms/grpo.py (hunk 1 above)
Comment (guyueh1) [reply nemo_rl/algorithms/grpo.py]: oh no, i think this is not intended, i will fix it

**Review by youngeunkwon0405 — DISMISSED** [review]
Great! Thanks for the contribution.

**Comment by guyueh1** [comment]
> @yfw @guyueh1 can we close out #2177 / #2178 and #2174 and just use this one?

#2177 is covered (I just pushed a new commit to cover a case unhandled by this PR).
#2178 is an optimization (to not initialize reference weights if no KL penalty), this is not covered by this PR, I feel it's not very necessary as it has no impact on perf. The impact is to save some initialization time and CPU memory. TLDR, This PR doesn't cover it.
#2174 should be covered by this PR too

File: nemo_rl/algorithms/grpo.py [hunk 2]
```diff
@@ -1727,10 +1725,26 @@ def grpo_train(
 
                     metrics_logging_data["content"] = flat_messages["content"]
 
+                force_on_policy_ratio = master_config["loss_fn"].get(
+                    "force_on_policy_ratio", False
+                )
```
[the remark this answers is not in the document]
Comment (guyueh1) [reply nemo_rl/algorithms/grpo.py]: this is ok to ignore

File: nemo_rl/algorithms/grpo.py [hunk 3]
```diff
@@ -1744,13 +1758,12 @@ def grpo_train(
                             **extra_multimodal_data,
                         }
```
[the remark this answers is not in the document]
Comment (guyueh1) [reply nemo_rl/algorithms/grpo.py:1780]: minor, could be ignored

File: nemo_rl/algorithms/grpo.py [hunk 4]
```diff
@@ -1727,10 +1725,26 @@ def grpo_train(
...
-                print("▶ Preparing for logprob inference...", flush=True)
-                with timer.time("logprob_inference_prep"):
-                    policy.prepare_for_lp_inference()
+                if not (skip_prev_logprobs and skip_reference_policy_logprobs):
+                    print("▶ Preparing for logprob inference...", flush=True)
+                    with timer.time("logprob_inference_prep"):
+                        policy.prepare_for_lp_inference()
 
                 print("▶ Computing logprobs...", flush=True)
                 with timer.time("policy_and_reference_logprobs"):
```
[the remark this answers is not in the document]
Comment (guyueh1) [reply nemo_rl/algorithms/grpo.py:1771]: i think it's ok to ignore

File: nemo_rl/algorithms/grpo.py [hunk 5]
```diff
@@ -1727,10 +1725,26 @@ def grpo_train(
 
                     metrics_logging_data["content"] = flat_messages["content"]
 
+                force_on_policy_ratio = master_config["loss_fn"].get(
+                    "force_on_policy_ratio", False
```
[the remark this answers is not in the document]
Comment (guyueh1) [reply nemo_rl/algorithms/grpo.py]: ok to ignore, i don't want it to return None

**Comment by guyueh1** [comment]
dup of https://github.com/NVIDIA-NeMo/RL/pull/2443/changes
